prereise/cli/data_sources.py

```diff
@@ -0,0 +1,130 @@
+
+from prereise.cli.helpers import validate_date, validate_file_path
+from prereise.gather.winddata.rap import rap
+
+
+class DataSource(metaclass=ABCMeta):
+    """Abstract class that defines a strict interface
+    that future data sources need to follow in order
+    to seamlessly integrate with the parser.
+
+    :param abc.ABCMeta metaclass: defaults to ABCMeta
+    """
+
+    @property
+    @abstractmethod
+    def command_name(self):
+        """String command name for argparse. See add_parser function
+        at `this link <https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_subparsers>`_.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def command_help(self):
+        """String with a help description for the command. See add_parser function
+        at `this link <https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_subparsers>`_
+        for more details
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def extract_arguments(self):
+        """List of dictionaries defining the various flags and parameters
+        the user needs to pass into this command, which are
+        subsequently passed to the below extract function.
+        See `add_argument <https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_argument>`_
+        to understand what is acceptable/unacceptable in these
+        dictionaries.
+        """
+        pass
+
+    @abstractmethod
+    def extract(self):
+        """Function that takes in the arguments defined in
+        the above property extract_arguments and downloads
+        the appropriate data and saves it somewhere.
+        """
+        pass
+
+
+class WindData(DataSource):
+    @property
+    def command_name(self):
+        """See :py:func:`prereise.cli.data_sources.DataSource.command_name`
+
+        :return: (*str*)
+        """
+        return "wind_data"
+
+    @property
+    def command_help(self):
+        """See :py:func:`prereise.cli.data_sources.DataSource.command_help`
+
+        :return: (*str*)
+        """
+        return "Download wind data"
+
+    @property
+    def extract_arguments(self):
+        """See :py:func:`prereise.cli.data_sources.DataSource.extract_arguments`
+
+        :return: (*str*)
+        """
+        return [
+            {
+                "command_flags": ["--region", "-r"],
+                "required": True,
+                "choices": ["Texas", "Eastern", "Western"],
+                "type": str,
+                "action": "append",
+            },
+            {
+                "command_flags": ["--start_date", "-sd"],
+                "required": True,
+                "type": validate_date,
+            },
+            {
+                "command_flags": ["--end_date", "-ed"],
+                "required": True,
+                "type": validate_date,
+            },
+            {
+                "command_flags": ["--file_path", "-fp"],
```

Comment:
Similar to the above comment, it would be nice to specify that this filepath needs to include the filename and that the resulting file will be a pickle file so the user can use the appropriate extension (to keep consistent with the rest of the codebase, we use `.pkl` as the pickle file extension).

Reply:
Done